JsonNumber implementation and tests

kasiabulat opened the pull request:
I implemented `JsonNumber` methods using string to keep the actual value (and added more tests for it). I also added more equals-related methods to JsonNumber, JsonBoolean and JsonString.

cc: @joperezr @bartonjs @ericstj @ahsonkhan @terrajobst

kasiabulat committed:
IEquatables added

kasiabulat committed:
JsonNUmber unit tests added, minor fixes

kasiabulat committed:
constructors for JsonNumber added

kasiabulat committed:
tests made public

kasiabulat committed:
tests fixes

kasiabulat committed:
constructors changed to use setters

kasiabulat committed:
implicit operators changed to return new JsonNumber

kasiabulat committed:
byte[] changed to string, try get methods implemented

kasiabulat committed:
equals-related methods added

kasiabulat committed:
more tests added

joperezr commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -90,14 +90,17 @@ public partial class JsonArray : System.Text.Json.JsonNode, System.Collections.G
         System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw null; }
     }
-    public partial class JsonBoolean : System.Text.Json.JsonNode
+    public partial class JsonBoolean : System.Text.Json.JsonNode, System.IEquatable<System.Text.Json.JsonBoolean>
```
I suppose these you changed only because you were already changing JsonNumber. For the main fork PR only send JsonNumber class, which I suppose is already your plan, just wanted to reiterate :)

joperezr commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -264,7 +267,7 @@ public abstract partial class JsonNode
         public static bool TryGetNode(System.Text.Json.JsonElement jsonElement, out System.Text.Json.JsonNode jsonNode) { throw null; }
     }
-    public partial class JsonNumber : System.Text.Json.JsonNode
+    public partial class JsonNumber : System.Text.Json.JsonNode, System.IEquatable<System.Text.Json.JsonNumber>
```
@bartonjs would it make sense to also make this implement IEquatable<int>, IEquatable<double>, etc.?

ahsonkhan replied:
> make this implement IEquatable, IEquatable, etc.?

Was one of them meant to be generic?

bartonjs replied:
Assuming that your IEquatables were generic over numeric value types:

Not really.

The equality for this is either very complicated, simple but maybe not useful.  JSON numbers are any legal value in `Q` (the set of rational (real) numbers).  If we're looking at this from a mathy approach then "3" "3e0", "30e-1", "3.0", "3.00", and the infinite (up to memory limits) other possible ways of being "3" are all equal.  I doubt the implementation is going to take that into account.  I expect that it's just "do the two implementations have the same string, internally?".

An `IEquatable<int>` implementation could be written that does TryParse on the field and matches against the input.  But is it useful?

As far as explicit guidelines, all we have are that immutable value types should be IEquatable of themselves, and you should consider being IEquatable of yourself if you override Object.Equals, and if you're IComparable of yourself you should be IEquatable of yourself.

We don't have one that says "don't be IEquatable of not-you", but instinctively (as a mathematician) that "do not" makes sense to me because the inverse relationship isn't true (int isn't equatable to JsonNumber, so JsonNumber shouldn't be equatable to int, by symmetry)

joperezr replied:
> Was one of them meant to be generic

Yes sorry, codeflow didn't like the fact that I didn't use code blocks to define my generics 😆

joperezr commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,218 @@
+
+namespace System.Text.Json
+{
+//#pragma warning disable xUnit1000
```
NIT: now that you made the class public, just remove these comments.

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
-        public long GetInt64() { throw null; }
-        public float GetSingle() { throw null; }
-        public double GetDouble() { throw null; }
+        public string GetString()
```
I'm not 100% about this one, shouldn't this be instead an override to ToString method? I get that in the others it totally makes sense to have the Get in order to get the value out, I just don't know how will users percieve a JsonNumber.GetString method, it is not trivial knowing that the way we store the value is with a string.

kasiabulat replied:
Sure, I'll change it.

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
+        }
+        public byte GetByte()
+        {
+            return byte.Parse(_value);
```
NIT: all these methods can be written instead as oneliners which I think makes the file look cleaner:
```c#
public byte GetByte() => byte.Parse(_value);
```

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
-        public bool TryGetInt64(out long value) { throw null; }
-        public bool TryGetSingle(out float value) { throw null; }
-        public bool TryGetDouble(out double value) { throw null; }
+        public bool TryGetByte(out byte value)
```
same for these, you could do:
```c#
public bool TryGetByte(out byte value) => byte.TryParse(_value, out value);
```

kasiabulat replied:
Oh yes, thanks! :)

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
+        }
+        public void SetDouble(double value)
+        {
+            _value = value.ToString();
```
question: will this roundtrip when using exponential notation?

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
 namespace System.Text.Json
 {
-    public partial class JsonNumber : JsonNode
+    public partial class JsonNumber : JsonNode, IEquatable<JsonNumber>
```
Would it also make sense to implement IComperable<JsonNumber>?

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
+           catch
+            {
+                value = 0;
+                return false;
```
Try methods shouldn't be written as catch/ignore/return-false.  Implementing a non-Try method in terms of the Try is legitimate, though (the question I thought you asked me in person earlier).

But in this case, as @joperezr pointed out, the ideal implementation is to just defer to byte.TryParse (et al).

ahsonkhan replied:
As a general rule, try not to use exceptions (try/catch) as control flow (especially a catch-all):
https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
+        {
+            SetString(value);
+        }
+        public JsonNumber(byte value)
```
nit: add spaces between methods

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
+            }
+        }
+
+        public void SetString(string value)
```
This probably needs a better name, like `SetFormattedValue`.

kasiabulat replied:
Ok, I'll change it.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
+        }
+        public void SetDouble(double value)
+        {
+            _value = value.ToString();
```
This probably wants a specific format (like G15).  Whatever Utf8JsonWriter does.

kasiabulat replied:
Ok, I checked it - "G15" is default for ToString and Utf8JsonWriter uses default StandardFormat - that is also "G", but with unspecified precision. Should I leave the default format or change it to "G17" or something else?

bartonjs replied:
I thought we had changed Utf8JsonWriter to specify a precision.  If not, nothing is needed here :smile:.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
-        public JsonNumber(double value) { }
+        public JsonNumber(string value)
+        {
+            SetString(value);
```
Explicitly validate and throw for null?
Unless you are implicitly relying on `Encoding.UTF8.GetBytes` to do the validation for you.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
     {
+        private string _value;
+
         public JsonNumber() { }
```
The default ctor here is the only time that `_value` can be `null`. Consider removing it, or initializing `_value` to `"0"`.

bartonjs commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,218 @@
+        [Fact]
+        public static void TestShort()
+        {
+            var jsonNumber = new JsonNumber();
```
The fact that this type has a default ctor surprised me.  There should be a test that tests the behaviors of this ctor (what throws, what succeeds, etc).

Either it should be 0 (in which case all methods work) or it should not exist.

ahsonkhan replied:
> The fact that this type has a default ctor surprised me.

Yep, same here.

kasiabulat replied:
Ok, I can change it to "0".

bartonjs commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,218 @@
+        public static void TestInt()
+        {
+            var jsonNumber = new JsonNumber();
+            int value = 17;
```
Missing tests:

* Things used the typed ctors
* Creating a value with one type, then changing it to another.
* When the Get methods fail / the TryGet methods return false.
* ToString/GetString/whatever should always succeed.

ahsonkhan replied:
Also implicit casts.

kasiabulat replied:
Thanks, I'll add those.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(_value);
```
nit: How about just `_value.GetHashCode()`?

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,218 @@
+        [InlineData("0.")]
+        [InlineData("008")]
+        [InlineData("5e")]
+        public static void TestInvalidString(string value)
```
We should consider expanding on the "invalid string" cases.

See these for some more:
https://github.com/dotnet/corefx/blob/9e994a5923d426f0e64ed3b8f9b93df6f214a42a/src/System.Text.Json/tests/Utf8JsonReaderTests.cs#L4382-L4393

Consider also including white space before/after.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,218 @@
+        public static void TestUInt64()
+        {
+            var jsonNumber = new JsonNumber();
+            ulong value = 5;
```
Consider adding more, especially ones that are within the uint64 range but not in int32/int64 ranges (like uint64.maxvalue).

It would be good to test the boundary values.

ahsonkhan replied:
Same goes for other overloads (including negative numbers).

kasiabulat replied:
Ok, sure!

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,218 @@
+            double value = 3.14;
+
+            jsonNumber.SetDouble(value);
+            double result = jsonNumber.GetDouble();
```
What do you get if you call `jsonNumber.GetInt32()` here? Is it valid to do so?

It might be good to structure the tests so that you can create a JsonNumber from some input (say a string) and then call all sorts of mismatching `GetX()` APIs on them.

So, SetX(some numeric type), GetX(another numeric type).

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,218 @@
+        public static void TestJsonNumberConstructor(object value)
+        {
+            // should not throw any exceptions:
+            var jsonNumber = new JsonNumber(value as dynamic);
```
I would consider adding to this test and showcasing the behavior of each {Try}GetX() API.
```suggestion
            var jsonNumber = new JsonNumber(value as dynamic);
            // call and assert jsonNumber.GetDouble();
            // call and assert jsonNumber.TryGetDouble();
            // call and assert jsonNumber.GetSingle();
            // ... etc.
```

ahsonkhan commented:
cc @JamesNK

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -7,89 +7,357 @@
 namespace System.Text.Json
 {
-    public partial class JsonNumber : JsonNode
+    public partial class JsonNumber : JsonNode, IEquatable<JsonNumber>
```
Presumably we are defering API documentation (`/// xml comments` like summary/parameters/returns/exceptions) until after the API is more stable. Is the plan to add them once the type/impl/tests moves to corefx?

kasiabulat replied:
I'll add documentation in main corefx PR.

ahsonkhan commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -283,6 +286,7 @@ public partial class JsonNumber : System.Text.Json.JsonNode
         [System.CLSCompliantAttribute(false)]
         public JsonNumber(ulong value) { }
```
What about decimal?

kasiabulat replied:
I'll add it.

JamesNK commented:
> cc @JamesNK

I don't have any context here. Could you point me at a spec/proposal/issue that discusses what these APIs are for

ahsonkhan commented:
> I don't have any context here. Could you point me at a spec/proposal/issue that discusses what these APIs are for

This is still in the initial prototype/design phase and maybe @joperezr / @kasiabulat have some doc/spec they can point you to (if we don't have one yet, it might be good to write one up at some point).

We are starting to fill in implementation details and then go through the API review/design in corefx within the next month or so (before @kasiabulat's internship ends), so there is no issue yet.

Essentially, we are looking to create a modifiable, dictionary-backed Json DOM API to complement the readonly `JsonDocument`/`JsonElement`, with type hierarchy consisting of the base type `JsonNode` (rather than a single union type `JsonElement`).
The user should be able to build up a structured in-memory representation of the JSON payload and modify it. That includes, remove, add, and update (similar to Newtonsoft.Json's `JToken`).

The `WritableJsonApiTests.cs` contains a bunch of usage scenarios that we want to enable with these types. For example:
https://github.com/kasiabulat/corefx/blob/25109c364e5bcb2d70a2a7b91f620e28faab27ca/src/System.Text.Json/tests/WritableJsonApiTests.cs#L19-L42

I cc'd you primarily to get your eyes on it, and to get your insights early (if applicable), especially to avoid common pitfalls or design mistakes that you may have encountered (and to get your general opinion on the design). If you see any big rocks or have significant concerns, it would be great to discuss and address them before we have gone too deep with the current approach.

kasiabulat committed:
pragmas deleted, methods changed into oneliners

kasiabulat committed:
review comments included, more tests added

kasiabulat commented:
I changed methods into one-liners and added more tests, but still need to add some for Get/TryGet for different types (correct and incorrect examples). Personally, I think it should be possible to GetInt after setting something that fits into int and not possible if it doesn't etc. And in current implementation I forget the information of what exact type was stored.

kasiabulat committed:
(Try)Get tests on different types added

kasiabulat committed:
ToString and changing types tests added

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -5,91 +5,161 @@
+        public void SetFormattedValue(string value)
+        {
+            if (value == null)
+                throw new NullReferenceException("Expected number, but instead got null.");
```
This should be `ArgumentNullException`. I believe we never explicitly throw NullRef (and its only thrown by the runtime). For example:
https://github.com/dotnet/corefx/blob/9d6729a940952dcc14dc8683218684f4da29198a/src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs#L93-L96

kasiabulat replied:
Ok, I'll change it.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -5,91 +5,161 @@
+        {
+            if (value == null)
+                throw new NullReferenceException("Expected number, but instead got null.");
+            if (value == "")
```
nit: Checking length will likely be much faster (and avoid a call to `.Equals`). [sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgMwAEu5AwuQN7HlOUXAD2bANuQLJ4AUWPAAZyANwicAlPUbN5WAOzjJ5ALxryAIi0BuOUwC+BluXZdeOQSJXTZReQuUTOAOgAycAHYBzGAAt1TWF9B2ZjIkMgA===) 
```suggestion
            if (value.Length == 0)
```

bartonjs replied:
Or combine both with `string.IsNullOrEmpty(value)`

kasiabulat replied:
I split those cases to throw different exceptions.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -5,91 +5,161 @@
+            if (value == "")
+                throw new ArgumentException("Expected number, but instead got empty string.");
+            
+            JsonWriterHelper.ValidateNumber(Encoding.UTF8.GetBytes(value).AsSpan());
```
Consider adding "throw argumentexception" validation to a copy of `JsonWriterHelper.ValidateNumber` (with Debug.Asserts changed to actual checks).

The length check can go there as well:
https://github.com/dotnet/corefx/blob/9d6729a940952dcc14dc8683218684f4da29198a/src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs#L137-L148

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -5,91 +5,161 @@
+                throw new ArgumentException("Expected number, but instead got empty string.");
+            
+            JsonWriterHelper.ValidateNumber(Encoding.UTF8.GetBytes(value).AsSpan());
+            _value = value;
```
I would suspect these strings to generally be small (say < 256 characters). Consider instead, stackalloc a span and use the `Span<byte>` overload of `Encoding.UTF8.GetBytes` to avoid the allocation on .NET Core 3.0 (and use the arraypool for larger length strings).
https://github.com/dotnet/corefx/blob/ef5b086fc5f134e3244d0bca10d70db8055fa1c6/src/System.Runtime/ref/System.Runtime.cs#L7683

OR

Validate the value as chars.

Something to consider to optimize the implementation a bit (but that can be deferred to later, if you like).

kasiabulat committed:
implicit operators test added

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -5,91 +5,161 @@
+        
+        public long GetInt64() => long.Parse(_value);
+        
+        public float GetSingle() => float.Parse(_value);
```
Why is `decimal` not supported here? I think we should support all the .NET types the existing `JsonDocument`/`JsonElement` supports

kasiabulat replied:
Yes, I plan to add it.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -5,91 +5,161 @@
+        
+        public float GetSingle() => float.Parse(_value);
+        
+        public double GetDouble() => double.Parse(_value);
```
**Question:** Would escaped characters be supported? Especially since the `Utf8JsonWriter` escapes the `+` character by default (since it uses `JavascriptEncoder.Default`).

```C#
Console.WriteLine(double.Parse("25.1e+1")); // valid - 251
Console.WriteLine(double.Parse("25.1e\\u002B1")); // Fails
```

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,519 @@
+            byte value = 17;
+            var jsonNumber = new JsonNumber(value);
+
+            // Geeting other types should also succeed:
```
nit: typo
```suggestion
            // Getting other types should also succeed:
```

Same elsewhere (everywhere else it was copied).

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,519 @@
+            // Geeting other types should also succeed:
+            short shortResult = jsonNumber.GetInt16();
+            Assert.Equal(value, shortResult);
+            bool success = jsonNumber.TryGetInt16(out shortResult);
```
nit: It doesn't look like the `success` local is used for anything meaningful.

How about just assert:
```C#
Assert.True(jsonNumber.TryGetInt16(out shortResult));
```

kasiabulat replied:
Sure, thanks!

kasiabulat committed:
code refactored

kasiabulat committed:
decimal support added to JsonNumber

kasiabulat committed:
decimal support added to JsonArray and JsonObject

kasiabulat commented:
I added decimal support and addressed other minor issues. There are still some questions left to discuss, but I am merging this PR and I will create one to the main corefx fork.

kasiabulat commented:
I created the issue for my project: https://github.com/dotnet/corefx/issues/39922
